Introduz fluxo de cadastro em 2-passos

berinhard opened the pull request:
Fixes #433 

Demo em vídeo [aqui](https://www.youtube.com/watch?v=OPTfBMroE2A).

No vídeo eu acabei de mostrar a conta que criei sendo apresentada na listagem da newsletter. Segue o print mostrando ela na listagem:

![Screenshot from 2020-09-28 18-47-55](https://user-images.githubusercontent.com/238223/94489599-23703380-01bb-11eb-9dfb-294b324691a5.png)

berinhard committed:
Instala e configure django-registration

berinhard committed:
Adiciona rotas para fluxo de autenticação de 2 passos

berinhard committed:
Cria templates dummy que são necessários

berinhard committed:
Configura views para funcionar de acordo com estrutura de rotas do brasil.io

berinhard committed:
Melhora cobertura de testes para garantir uso correto da lib

berinhard committed:
Move ativação de newsletter para view porque user não é salvo no form durante o novo processo

berinhard committed:
Já loga a usuária após validá-la

berinhard committed:
Make lint

berinhard committed:
Adiciona conteúdo as novas páginas

turicas commented on brasilio/settings.py:
```diff
@@ -280,3 +281,8 @@ def get_neo4j_config_dict(neo4j_uri):
 CLOUDFLARE_BLOCKED_IPS_RULE = env("CLOUDFLARE_BLOCKED_IPS_RULE")
+
+# Django registration config
+ACCOUNT_ACTIVATION_DAYS = env("ACCOUNT_ACTIVATION_DAYS", cast=int)
+REGISTRATION_SALT = env("REGISTRATION_SALT")
+REGISTRATION_OPEN = env("REGISTRATION_OPEN", cast=bool)
```
Acho que poderiam ter valores padrão (`7`, `SECRET_KEY`, `True`).

berinhard replied:
Resolvido em 
2226fca

turicas commented on brasilio_auth/templates/brasilio_auth/activation_complete.html:
```diff
@@ -1,5 +1,10 @@
+  <div class="row m-t-15 center">
+    <div class="col s12 m8 p8 offset-m2">
+      <h4>Ativação realizada com sucesso!</h4>
+      <p>Pronto! Agora você está liberada para acessar todas as páginas do Brasil.io</p>
```
```suggestion
      <p>
        Pronto! Sua conta foi ativada e agora você pode acessar todas as páginas do <a href="https://brasil.io/">Brasil.IO</a> (incluindo nossa API). :-)
      </p>
      <p>
        Caso você precise baixar quantidades grandes de dados, por favor opte por <b>baixar os dados completos de cada dataset</b> em vez de utilizar a API ou acessar diretamente a listagem dos dados (basta acessar a página do dataset e clicar em "baixar dados completos").
      </p>
      <p>
        O <a href="https://brasil.io/">Brasil.IO</a> é <b>desenvolvido colaborativamente por voluntários</b>, fica aqui o convite para <a href="{% url "core:collaborate" %}">colaborar com o projeto</a>. :-) Considere também <a href="{% url "core:donate" %}">fazer-nos uma doação</a>.
      </p>
```

turicas commented on brasilio_auth/templates/brasilio_auth/emails/activation_email_body.txt:
```diff
@@ -1 +1,11 @@
-corpo do email de ativação
+Olá, {{ user.username }}!
+
+Obrigado por se cadastrar no Brasil.io =)
```
```suggestion
Obrigado por se cadastrar no Brasil.IO! =)
```

berinhard replied:
Atualiza typo no corpo do email de ativação de senha

turicas commented on brasilio_auth/templates/brasilio_auth/emails/activation_email_subject.txt:
```diff
@@ -1 +1 @@
-assunto do email de ativação
+[Brasil.io] Ative sua conta
```
```suggestion
[Brasil.IO] Ative sua conta
```

turicas commented on brasilio_auth/templates/brasilio_auth/registration_closed.html:
```diff
@@ -1,5 +1,16 @@
+        Infelizmente desativamos temporáriamente o cadastro de novas pessoas por questões técnicas.
+      </p>
+      <p>
+        Fique atento ao nosso <a href="https://t.me/brasil_io">grupo do Telegram</a> que é sempre a principal fonte de atualizações sobre o Brasil.io e suas atualizações.
```
```suggestion
        Acompanhe nosso <a href="https://t.me/brasil_io">canal no Telegram</a> para saber mais novidades.
```

turicas commented on brasilio_auth/templates/brasilio_auth/registration_closed.html:
```diff
@@ -1,5 +1,16 @@
-sistema fechado para cadastro de novos usuários
+  <div class="row m-t-15 center">
+    <div class="col s12 m8 p8 offset-m2">
+      <h4>Ops, estamos fechados para novos cadastros</h4>
```
```suggestion
      <h4>Ops, estamos fechados para novos cadastros!</h4>
```

turicas commented on brasilio_auth/templates/brasilio_auth/registration_closed.html:
```diff
@@ -1,5 +1,16 @@
+    <div class="col s12 m8 p8 offset-m2">
+      <h4>Ops, estamos fechados para novos cadastros</h4>
+      <p>
+        Infelizmente desativamos temporáriamente o cadastro de novas pessoas por questões técnicas.
```
```suggestion
        Desativamos temporariamente novos cadastros por questões técnicas.
```

turicas commented on brasilio_auth/templates/brasilio_auth/registration_complete.html:
```diff
@@ -1,5 +1,16 @@
-cadastro de usuário com sucesso
+  <div class="row m-t-15 center">
+    <div class="col s12 m8 p8 offset-m2">
+      <h4>Dê uma olhadinha no seu email...</h4>
+      <p>
+        Enviamos um email para realizar a confirmação do seu cadastro no Brasil.io.
+      </p>
+      <p>
+        Abra o email e clique no link de ativação para finalizar o seu cadastro com sucesso.
+      </p>
+
```
```suggestion
      <h4>Verifique seu e-mail!</h4>
      <p>
        Enviamos um e-mail com o link para ativar seu cadastro. Acesse seu e-mail e clique no link para começar a desfrutar de tudo que o <a href="https://brasil.io/">Brasil.IO</a> pode te oferecer. :-)
      </p>
      <p>
        O <a href="https://brasil.io/">Brasil.IO</a> é <b>desenvolvido colaborativamente por voluntários</b>, fica aqui o convite para <a href="{% url "core:collaborate" %}">colaborar com o projeto</a>. :-) Considere também <a href="{% url "core:donate" %}">fazer-nos uma doação</a>.
      </p>
```

turicas commented on brasilio_auth/forms.py:
```diff
@@ -1,13 +1,13 @@
 
 
-class UserCreationForm(DjangoUserCreationForm):
+class UserCreationForm(RegistrationFormUniqueEmail):
```
Dúvida: essa alteração muda em alguma coisa aquela verificação de username único (que tínhamos feito para caso tenhamos cadastrado "Username" e tentem cadastrar "username" não tenhamos problemas)?

turicas replied:
Ficou faltando só responder essa.

berinhard replied:
Não, não muda nada. Tem [teste unitário](https://github.com/turicas/brasil.io/blob/develop/brasilio_auth/tests/test_forms.py#L100) para isso já que continua passando, inclusive.

turicas commented on brasilio_auth/urls.py:
```diff
@@ -41,10 +40,14 @@
-    path("entrar/", disable_non_logged_user_cache(CreateUserView.as_view()), name="sign_up"),
+    path(
+        "entrar/",
+        RegistrationView.as_view(),
```
O `disable_non_logged_user_cache` não é necessário aqui?

berinhard replied:
Resolvido em aa3bc2b

turicas commented on brasilio_auth/tests/test_views.py:
```diff
@@ -50,6 +51,16 @@ def test_create_inactive_user_and_redirect_to_sign_up_complete(self):
         self.assertRedirects(response, reverse("brasilio_auth:sign_up_complete"))
+        assert not NewsletterSubscriber.objects.exists()
+
+    def test_create_user_as_newsletter_subscriber(self):
+        self.data["subscribe_newsletter"] = True
+
+        self.client.post(self.url, data=self.data)
+
+        user = User.objects.get(username="foo")
+        assert not user.is_active
+        assert NewsletterSubscriber.objects.filter(user=user).exists()
```
Dúvida: o email do usuário é cadastrado na newsletter apenas depois que ele ativar o cadastro? Acredito que esse deva ser o fluxo correto (assim não enviaremos email de newsletter a quem se cadastrou, marcou a checkbox mas não confirmou o email).

berinhard replied:
No caso, não, o newsletter é criado na hora. Fazer dessa forma vai ser meio rolê porque, em algum lugar, vamos ter que guardar que a pessoa optou pela newsletter e ler essa info após a ativação.

Vou resolver isso criando um manager como `Newsletter.objects.active()` e atualizar o processo de download do CSV dos assinantes para usá-lo.

berinhard replied:
Resolvido em 4f555c4

berinhard committed:
Adiciona valores padrões para a app django-registration

berinhard committed:
Atualiza template de cadastro completo

Co-authored-by: Álvaro Justen <alvarojusten@gmail.com>

berinhard committed:
Atualiza typo no corpo do email de ativação de senha

Co-authored-by: Álvaro Justen <alvarojusten@gmail.com>

berinhard committed:
Atualiza typo no título do email de ativação de senha

Co-authored-by: Álvaro Justen <alvarojusten@gmail.com>

berinhard committed:
Melhora mensagem com link para o telegram

Co-authored-by: Álvaro Justen <alvarojusten@gmail.com>

berinhard committed:
Melhora mensagem no caso de cadastro fechado

Co-authored-by: Álvaro Justen <alvarojusten@gmail.com>

berinhard committed:
Melhora mensagem após início de novo cadastro

Co-authored-by: Álvaro Justen <alvarojusten@gmail.com>

berinhard committed:
Desativa cache para página de criação de user

berinhard committed:
Adiciona arquivos __init__ que faltavam em módulos de testes

Por conta disso, o pytest gera conflito porque considera o test_models
de apps diferentes o mesmo arquivo

berinhard committed:
Considera somente usuárias ativas para envio da newsletter

berinhard commented:
@turicas já apliquei todas as alterações do HTML que você sugeriu e também arrumei os outros pontos. Para facilitar o review, tô mandando prints de todos as páginais com o conteúdo atual:

**Fechado para novos cadastros**
![Screenshot from 2020-09-29 15-34-39](https://user-images.githubusercontent.com/238223/94601310-b23b8980-0269-11eb-8e27-f84aabd8c357.png)

**Solicitação para verificar o email**
![Screenshot from 2020-09-29 15-35-37](https://user-images.githubusercontent.com/238223/94601349-bff10f00-0269-11eb-8822-e9669dbcf4ef.png)

**Email enviado**
![Screenshot from 2020-09-29 15-35-46](https://user-images.githubusercontent.com/238223/94601380-c97a7700-0269-11eb-886c-fb356e849f8b.png)

**Ativação finalizada**
![Screenshot from 2020-09-29 15-36-10](https://user-images.githubusercontent.com/238223/94601399-d1d2b200-0269-11eb-9571-a57d1c630c8a.png)

**Erro na ativação**
![Screenshot from 2020-09-29 15-36-27](https://user-images.githubusercontent.com/238223/94601412-d7c89300-0269-11eb-8c72-30e5ba927fc9.png)

turicas reviewed (approved):
Muito bom! Ficou faltando só responder a questão da verificação do username, mas se estiver ainda sendo checado, pode fazer o merge e deploy. :)

berinhard commented:
Show vou mergear, fazer o deploy e testar o fluxo pra ver se tá rolando mesmo =)
